tactic/subtype_instance.lean

```diff
@@ -0,0 +1,77 @@
+/-
+Copyright (c) 2018 Simon Hudon. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Simon Hudon
+
+Provides a `subtype_instance` tactic which builds instances for algebraic substructures
+(sub-groups, sub-rings...).
+-/
+
+import data.string
+import tactic.interactive tactic.algebra
+import data.subtype data.set.basic
+open tactic expr name list
+
+section tactic
+
+open tactic.interactive (get_current_field refine_struct)
+open lean lean.parser
+open interactive
+
+/-- make names as in `is_add_subgroup` from `add_group` by global prefacing with `is_` 
+    and prefacing the last underscore delimeted item with `sub` -/
+def mk_substructure_name (x : string) : string :=
+"is_" ++ string.map_tokens '_' (list.map_last $ (++) "sub") x
```

Comment:
A better way is analyse what the substructure is by looking into the local context in `subtype_instance`.

We have the following shape:
```lean
instance subset.ring {S : set R} [is_subring S] : ring S := by subtype_instance
```
It should be easy to figure out that `is_subring S` is the correct instance and structure name.

Reply:
I'm not sure what you're getting at. `mk_substructure_name` and `mk_mem_name` are used to take `monoid.mul` and turn it into `is_submonoid.mul_mem`. This is harder to get at using assumptions. I was considering using attributes but so far, the naming convention was the only thing I managed to use.